### PR TITLE
Improve errors

### LINKED ISSUE #154: Improve error debuggability
if i make a mistake updating a Template that in fact does not exist, i only see:

```
HTTPError: Response code 400 (Bad Request)
    at Request.<anonymous> (node_modules/got/dist/source/as-promise/index.js:118:42)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'ERR_NON_2XX_3XX_RESPONSE',
  timings: {
    start: 1671710227500,
    socket: 1671710227501,
    lookup: 1671710227501,
    connect: 1671710227537,
    secureConnect: 1671710227581,
    upload: 1671710227582,
    response: 1671710227789,
    end: 1671710227790,
    error: undefined,
    abort: undefined,
    phases: {
      wait: 1,
      dns: 0,
      tcp: 36,
      tls: 44,
      request: 1,
      firstByte: 207,
      download: 1,
      total: 290
    }
  },
  transloaditErrorCode: 'TEMPLATE_VALIDATION_ERROR'
}
```

I don’t know why it failed.

Yet if i code `node_modules/transloadit/` to manually dump the body when this happens, I can see the API does reveal more about the nature of my fail:

```js
{
  body: {
    error: 'TEMPLATE_VALIDATION_ERROR',
    message: "Invalid id/name: A Template with id/name 'xyz' doesn't exist for this app.",
    http_code: 400
  }
}
```

TLDR:

Would be cool to add `body.message` to the error that the consuming dev sees. Also `got` [swallows async stack traces](https://github.com/sindresorhus/got/blob/main/documentation/async-stack-traces.md). we might want to move off `got`. See #199

### PR DESCRIPTION
**Pull request by mifi**
Renamed `TransloaditError` to `ApiError`. Differences between `TransloaditError` and `ApiError`:
- Moved `TransloaditError.response.body` to `ApiError.response`
- Removed `TransloaditError.assemblyId` (can now be found in `ApiError.response.assembly_id`
- Removed `TransloaditError.transloaditErrorCode` (can now be found in `ApiError.response.error`
- `ApiError` does not inherit from `got.HTTPError`, but `ApiError.cause` will be the `got.HTTPError` instance that caused this error (except for when Tranloadit API responds with HTTP 200 and `error` prop set in JSON response, in which case cause will be `undefined`).

Note that (just like before) when the Transloadit API responds with an error we will always throw a `ApiError` - In all other cases (like request timeout, connection error, TypeError etc.), we don't wrap the error in `ApiError`.

Also improved error stack traces, added a unit test in `mock-http.test.ts` that verifies the stack trace.

See which implementation you like more, this or #212

closes #154

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -419,31 +415,35 @@ const url = client.getSignedSmartCDNUrl({
...
+Any errors originating from Node.js will be passed on and we use [GOT](https://github.com/sindresorhus/got) v11 for HTTP requests. [Errors from `got`](https://github.com/sindresorhus/got/tree/v11.8.6?tab=readme-ov-file#errors) will also be passed on, _except_ the `got.HTTPError` which will be replaced with a `transloadit.ApiError`, which will have its `cause` property set to the instance of the original `got.HTTPError`. `transloadit.ApiError` has these properties:
 
-- **(deprecated: use `cause` instead)** `HTTPError.response?.body` the JSON object returned by the server along with the error response (**note**: `HTTPError.response` will be `undefined` for non-server errors)
-- **(deprecated)** `HTTPError.transloaditErrorCode` alias for `HTTPError.cause?.error` ([View all error codes](https://transloadit.com/docs/api/response-codes/#error-codes))
-- `HTTPError.assemblyId` (alias for `HTTPError.response.body.assembly_id`, if the request regards an [Assembly](https://transloadit.com/docs/api/assemblies-assembly-id-get/))
+- `HTTPError.response` the JSON object returned by the server. It has these properties
```
Comment (mifi) [inline README.md:420]:
```suggestion
```

Comment (mifi) [inline README.md:420]:
```suggestion
- `response` the JSON object returned by the server. It has these properties:
```

**Comment by remcohaszing** [comment]
I suggest we go with #212 instead.

**Comment by mifi** [comment]
ok nice. please review this PR also, because #212 branches off of this PR

File: src/ApiError.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      const indexOfMessageEnd = cause.stack.indexOf(cause.message) + cause.message.length
+      const gotStacktrace = cause.stack.slice(indexOfMessageEnd)
+      this.stack = `${message}${gotStacktrace}`
+    }
+
+    // If we have an original stack, append it to the bottom, because `got`s stack traces are not very good
+    if (this.stack != null && appendStack != null) {
+      this.stack += `\n${appendStack.replace(/^([^\n]+\n)/, '')}`
+    }
+
```
Comment (remcohaszing) [inline src/ApiError.ts:46]: I think we should remove all of this. This is quite a lot of magic to avoid standard JavaScript behaviour. If the cause has a stack, it will be printed.

```suggestion

    // if we have a cause, use the stack trace from it instead
    if (cause != null && typeof cause.stack === 'string') {
      const indexOfMessageEnd = cause.stack.indexOf(cause.message) + cause.message.length
      const gotStacktrace = cause.stack.slice(indexOfMessageEnd)
      this.stack = `${message}${gotStacktrace}`
    }

    // If we have an original stack, append it to the bottom, because `got`s stack traces are not very good
    if (this.stack != null && appendStack != null) {
      this.stack += `\n${appendStack.replace(/^([^\n]+\n)/, '')}`
    }

```

File: test/unit/mock-http.test.ts [hunk 3]
```diff
@@ -92,39 +94,92 @@ describe('Mocked API tests', () => {
...
-        cause: expect.objectContaining({ assembly_id: '123' }),
+          'API error (HTTP 400) INVALID_FILE_META_DATA: Invalid file metadata https://api2-oltu.transloadit.com/assemblies/foo',
+        response: expect.objectContaining({ assembly_id: '123' }),
       })
     )
+
+    try {
+      await promise
+    } catch (err) {
+      expect(inspect(err).split('\n')).toEqual([
```
Comment (remcohaszing) [inline test/unit/mock-http.test.ts:132]: Nice, this is a clever way to assert multiline strings that don’t have to match exactly.

File: test/unit/mock-http.test.ts [hunk 4]
```diff
@@ -92,39 +94,92 @@ describe('Mocked API tests', () => {
...
-          body: expect.objectContaining({ assembly_id: '123' }),
-        }),
-        cause: expect.objectContaining({ assembly_id: '123' }),
+          'API error (HTTP 400) INVALID_FILE_META_DATA: Invalid file metadata https://api2-oltu.transloadit.com/assemblies/foo',
+        response: expect.objectContaining({ assembly_id: '123' }),
       })
     )
+
+    try {
+      await promise
```
Comment (remcohaszing) [inline test/unit/mock-http.test.ts:130]: Awaiting a something inside a try in a test to assert inside a catch, is pretty unsafe. What if the promise doesn’t throw? I see that’s asserted above, but it’s not immediately obvious the assertion above is a requirement for this assertion.

Testing errors can be tricky, and in my experience testing frameworks all tend so have different semantics around this. An idea:

```ts
const errorString = await promise.catch(inspect)
expect(errorString.split('\n')).toEqual([
  // …
])
```

File: src/Transloadit.ts [hunk 5]
```diff
@@ -114,14 +60,14 @@ function getHrTimeMs(): number {
...
-  // This happens sometimes when createAssembly with an invalid file (IMPORT_FILE_ERROR)
+  // This happens sometimes, for example when createAssembly with an invalid file (IMPORT_FILE_ERROR)
   if (
     typeof result === 'object' &&
     result !== null &&
     'error' in result &&
     typeof result.error === 'string'
   ) {
-    throw makeTransloaditError(new Error('Error in response'), result)
+    throw new ApiError({ body: result }) // in this case there is no `cause` because we don't have a HTTPError
```
Comment (remcohaszing) [inline src/Transloadit.ts:70]: Nitpicky grammar issue:

```suggestion
    throw new ApiError({ body: result }) // in this case there is no `cause` because we don't have an HTTPError
```

**Review by remcohaszing — CHANGES_REQUESTED** [review]

**Comment by mifi** [comment]
I agree with all your suggestions. I've added them to the other PR #212

**Comment by kvz** [comment]
Do i understand we first need to merge [this](https://github.com/transloadit/node-sdk/pull/211) and then [that](https://github.com/transloadit/node-sdk/pull/212)?

**Comment by mifi** [comment]
doesn't really matter i think

**Comment by mifi** [comment]
if we merge #212, this PR will contain that pr's commits too

**Comment by kvz** [comment]
Alright! Closing in favor of https://github.com/transloadit/node-sdk/pull/212

**Comment by mifi** [comment]
oops what i meant is that yes both PRs need to be merged but the order doesn't matter

**Comment by mifi** [comment]
i assume your intention was to merge this PR also then, so i'll do that now :)
